mongoclient/db.go

```diff
@@ -38,62 +38,29 @@ type Config struct {
 	SecondaryPreferred bool
 }
 
-// Connect connects to MongoDB
-func Connect(config *Config, log *logrus.Entry) (*mongo.Client, error) {
-	opts := options.Client().
-		SetConnectTimeout(config.ConnTimeout).
-		SetReplicaSet(config.ReplSetName).
-		SetHosts(config.Servers)
-
-	if config.AppName != "" {
-		opts.SetAppName(config.AppName)
-	}
-
-	if config.SecondaryPreferred {
-		opts.SetReadPreference(readpref.SecondaryPreferred())
-	}
-
-	if config.TLS != nil && config.TLS.Enabled {
-		tlsLog := log.WithFields(logrus.Fields{
-			"cert_file": config.TLS.CertFile,
-			"key_file":  config.TLS.KeyFile,
-			"ca_files":  strings.Join(config.TLS.CAFiles, ","),
-		})
-
-		tlsLog.Debug("Using TLS config")
-		tlsConfig, err := config.TLS.TLSConfig()
-		if err != nil {
+func ConnectWithOptions(log logrus.FieldLogger, replSet string, servers []string, opts ...Option) (*mongo.Client, error) {
+	finalOpts := options.Client().
+		SetConnectTimeout(time.Minute). // use a default
+		SetReplicaSet(replSet).
+		SetHosts(servers)
+	for _, opt := range opts {
+		if err := opt(finalOpts); err != nil {
 			return nil, err
 		}
-
-		opts.SetTLSConfig(tlsConfig)
-	} else {
-		log.Debug("Skipping TLS config")
 	}
-
-	if config.Auth != nil {
-		creds := options.Credential{
-			Username:   config.Auth.Username,
-			Password:   config.Auth.Password,
-			AuthSource: config.Auth.Source,
-		}
-
-		opts.SetAuth(creds)
-	}
-
 	log.WithFields(logrus.Fields{
-		"servers":     strings.Join(opts.Hosts, ","),
-		"replica_set": config.ReplSetName,
+		"servers":     strings.Join(finalOpts.Hosts, ","),
+		"replica_set": finalOpts.ReplicaSet,
 	}).Debug("Dialing database")
 
-	client, err := mongo.Connect(context.Background(), opts)
+	client, err := mongo.Connect(context.Background(), finalOpts)
 	if err != nil {
 		return nil, err
 	}
 
 	// Connect does not block for server discovery, so we should ping to ensure
 	// we are actually connected
-	ctx, cancel := context.WithTimeout(context.Background(), config.ConnTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), *finalOpts.ConnectTimeout+time.Second)
```

Comment:
Why the additional second?

Reply:
So I was thinking about this. We set the `SetConnectTimeout` and then we use a timeout on the context with the exact same. I think that this could lead to some inconsistency in the errors. For instance, if the context times out vs the underlying lib timing out.

I guess this raises the question to me, do we need a timeout enforced here? Or should we rely on the timeout from the library?

Reply:
This might be total request time vs. just a connect timeout. Thinking about it some more, we should probably separate the two timeouts.

Reply:
If we run into this being a request or issue i think we can surface it then